Figgle.Generator/RenderTextSourceGenerator.cs

```diff
@@ -97,244 +100,263 @@ public GenerateFiggleTextAttribute(string memberName, string fontName, string so
+            // for a single type, so that we can generate one file per type.
+            var typeToGenerateGroup = generationInfos.GroupBy(
+                keySelector: info => info.TargetType,
+                elementSelector: info => info.AttributeInfos,
+                comparer: SymbolEqualityComparer.Default);
 
-                if (type.Namespace is not null)
+            foreach (var generateGroup in typeToGenerateGroup)
+            {
+                var targetType = (ITypeSymbol)generateGroup.Key!;
+
+                // There may be multiple partial definitions for the same symbol,
+                // each listing different attributes that may or may not repeat.
+                // By merging all attributes into a single hash set, we ensure there
+                // are no repeats and we generate all source in one file per type.
+                var attributeInfos = new HashSet<RenderItem>(RenderItemComparer.Instance);
+                foreach (var attributeInfo in generateGroup)
                 {
-                    sb.AppendLine(
-                        $$"""
-                        namespace {{type.Namespace}}
-                        {
-                        """);
+                    attributeInfos.UnionWith(attributeInfo);
                 }
 
-                sb.AppendLine(
-                    $$"""
-                        partial class {{type.Class}}
-                        {
-                    """);
+                if (!IsValidTypeForGeneration(context, targetType))
+                {
+                    return;
+                }
+
+                var renderInfoBuilder = ImmutableArray.CreateBuilder<RenderSourceInfo>(
+                    attributeInfos.Count);
 
-                foreach (var item in data.Items)
+                var memberNames = new HashSet<string>();
+                foreach (var generateFiggleInfo in attributeInfos)
                 {
-                    var font = FiggleFonts.TryGetByName(item.FontName)
-                        ?? TryParseFromAdditionalFiles(context, item.FontName);
+                    if (!SyntaxFacts.IsValidIdentifier(generateFiggleInfo.MemberName))
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(
+                            InvalidMemberNameDiagnostic,
+                            generateFiggleInfo.AttributeLocation ?? targetType.Locations[0],
+                            generateFiggleInfo.MemberName ?? "unknown"));
+                        continue;
+                    }
 
-                    if (font is null)
+                    if (!memberNames.Add(generateFiggleInfo.MemberName!))
                     {
                         context.ReportDiagnostic(Diagnostic.Create(
-                            UnknownFontNameDiagnostic,
-                            item.FontNameLocation,
-                            item.FontName));
+                            DuplicateMemberNameDiagnostic,
```

Comment:
Above you mention that names are now de-duped, yet this can still raise a diagnostic. What's the distinction?

Reply:
No, that makes sense. Thanks for clearing it up. It's fine in this state, and you mentioned that it simplifies the implementation, so all good.